data/src/main/java/com/linkedin/data/schema/compatibility/AnnotationCompatibilityChecker.java

```diff
@@ -0,0 +1,193 @@
+          Map<String, Object> newerResolvedProperties = newerCheckContextAndResolvedProperty.getValue();
+
+          // If olderResolvedProperties or newerResolvedProperties contains the same namespace as the provided SchemaAnnotationHandler,
+          // we do the annotation check
+          if (olderResolvedProperties.containsKey(annotationNamespace) || newerResolvedProperties.containsKey(
+              annotationNamespace))
+          {
+            AnnotationCompatibilityResult result =
+                handler.annotationCompatibilityCheck(olderResolvedProperties, newerResolvedProperties,
+                    olderCheckContextAndResolvedProperty.getKey(), newerCheckContextAndResolvedProperty.getKey());
+            if (!result.isCheckAnnotation())
+            {
+              _log.info("There is no implementation of annotationCompatibilityCheck in SchemaAnnotationHandler: "
+                  + handler.getClass().getName());
+              break;
+            }
+            results.add(result);
+          }
+          newerResolvedPropertiesMap.remove(pathSpec);
+        }
+        else
+        {
+          if (olderResolvedProperties.containsKey(annotationNamespace))
+          {
+            // olderResolvedPropertiesMap has a PathSpec which the newResolvedPropertiesMap does not have,
+            // it could be either newer schema removed annotation or the field is not existed, pass
+            // an empty newerResolvedPropertiesMap and empty newerAnnotationContext to the annotation check.
+            AnnotationCompatibilityResult result =
+                handler.annotationCompatibilityCheck(olderResolvedProperties, new HashMap<>(),
+                    olderCheckContextAndResolvedProperty.getKey(), new SchemaAnnotationHandler.AnnotationCheckContext());
+            if (!result.isCheckAnnotation())
+            {
+              _log.info("There is no implementation of annotationCompatibilityCheck in SchemaAnnotationHandler: "
+                  + handler.getClass().getName());
+              break;
+            }
+            results.add(result);
+          }
+        }
+      }
+      for (Map.Entry<PathSpec, Pair<SchemaAnnotationHandler.AnnotationCheckContext, Map<String, Object>>> e : newerResolvedPropertiesMap
+          .entrySet())
+      {
+        Pair<SchemaAnnotationHandler.AnnotationCheckContext, Map<String, Object>>
+            newerCheckContextAndResolvedProperty = newerResolvedPropertiesMap.get(e.getKey());
+        Map<String, Object> newerResolvedProperties = newerCheckContextAndResolvedProperty.getValue();
+        if (newerResolvedProperties.containsKey(annotationNamespace))
+        {
+          // newerResolvedPropertiesMap has a PathSpec which the olderResolvedPropertiesMap does not have,
+          // it could be either newer schema add new annotations or adding new field with annotations, pass
+          // an empty olderResolvedPropertiesMap and empty olderAnnotationContext to the annotation check.
```

Comment:
How would the compatibility checker differentiate between a new field(compatible) vs adding new annotations(may not be compatible)?

I think we should invoke the checker only for new annotations. And for new fields skip annotation checking? 
Or, send empty context/map for the case where new annotations were added. and use null when a new field was added.

Reply:
Once we return empty resolvedProperties in AnnotationCheckResolvedPropertiesVistor. We would figure out adding new annotation on an existing field use case. Older schema and newer schema will have the same path specs. It will be handled in the previous check.
 
For this branch, only new field with new annotation use case will fall in to this check. 

My question is for new field with new annotation case, do we need to do the compatibility check?

Reply:
My intuition says we can ignore the new field with new annotation case. Confirm with @junchuanwang as he has better context on annotations.

Reply:
Discussed offline with @nickibi, I recommend to add this check for new field with new annotation

Reply:
Add check for new field with new annotation case.